pallets/liquidity-rewards/src/lib.rs

```diff
@@ -0,0 +1,412 @@
+pub struct EpochChanges<BlockNumber, Balance, GroupId, CurrencyId, Weight, MaxChangesPerEpoch>
+where
+	MaxChangesPerEpoch: Get<u32>,
+	GroupId: Ord,
+	CurrencyId: Ord,
+{
+	duration: Option<BlockNumber>,
+	reward: Option<Balance>,
+	weights: BoundedBTreeMap<GroupId, Weight, MaxChangesPerEpoch>,
+	currencies: BoundedBTreeMap<CurrencyId, GroupId, MaxChangesPerEpoch>,
+}
+
+impl<BlockNumber, Balance, GroupId, CurrencyId, Weight, MaxChangesPerEpoch> Default
+	for EpochChanges<BlockNumber, Balance, GroupId, CurrencyId, Weight, MaxChangesPerEpoch>
+where
+	MaxChangesPerEpoch: Get<u32>,
+	GroupId: Ord,
+	CurrencyId: Ord,
+{
+	fn default() -> Self {
+		Self {
+			duration: None,
+			reward: None,
+			weights: BoundedBTreeMap::default(),
+			currencies: BoundedBTreeMap::default(),
+		}
+	}
+}
+
+pub type DomainIdOf<T> = <<T as Config>::Domain as TypedGet>::Type;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Required origin for admin purposes for configuring groups and currencies.
+		type AdminOrigin: EnsureOrigin<Self::Origin>;
+
+		/// Type used to handle balances.
+		type Balance: Balance + MaxEncodedLen + FixedPointOperand;
+
+		/// Domain identification used by this pallet
+		type Domain: TypedGet;
+
+		/// Type used to identify currencies.
+		type CurrencyId: AssetId + MaxEncodedLen + Clone + Ord;
+
+		/// Type used to identify groups.
+		type GroupId: Parameter + MaxEncodedLen + Ord + Copy;
+
+		/// Type used to handle group weights.
+		type Weight: Parameter + MaxEncodedLen + EnsureAdd + Unsigned + FixedPointOperand + Default;
+
+		/// The reward system used.
+		type Rewards: GroupRewards<Balance = Self::Balance, GroupId = Self::GroupId>
+			+ AccountRewards<
+				Self::AccountId,
+				Balance = Self::Balance,
+				CurrencyId = (DomainIdOf<Self>, Self::CurrencyId),
+			> + CurrencyGroupChange<
+				GroupId = Self::GroupId,
+				CurrencyId = (DomainIdOf<Self>, Self::CurrencyId),
+			> + DistributedRewards<Balance = Self::Balance, GroupId = Self::GroupId>;
+
+		/// Max groups used by this pallet.
+		/// If this limit is reached, the exceeded groups are either not computed and not stored.
+		#[pallet::constant]
+		type MaxGroups: Get<u32> + TypeInfo;
+
+		/// Max number of changes of the same type enqueued to apply in the next epoch.
+		/// Max calls to [`Pallet::set_group_weight()`] or to [`Pallet::set_currency_group()`] with
+		/// the same id.
+		#[pallet::constant]
+		type MaxChangesPerEpoch: Get<u32> + TypeInfo + sp_std::fmt::Debug + Clone + PartialEq;
+
+		/// Information of runtime weights
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	pub(super) type EndOfEpoch<T: Config> = StorageValue<
```

Comment:
Maybe we can add a comment explaining that we took the endOfEpoch out of the active epoch data storage to not have to deserialize the whole struct for 99% of the time assuming a duration of 100 blocks :)

Reply:
I like it!